### DISCUSSION TITLE
Question: Scroll text up to a clear screen?

### DISCUSSION THREAD

**Discussion by vefatica** [microsoft/terminal#13412] [Q&A] [answered]
In tcsh or bash, in WT or in a console, Ctrl-L scrolls text upwards so that the current command line (even one in progress) is at the top of the viewport ... without clobbering any previous output.  I'd like to implement this in TCC (where plugins may have an input keystroke handler}.  I've tried the scroll sequence (ESC[nS) but that clobbers whatever previous output is currently in the viewport.  How do tcsh and bash do that?  Any tips would be appreciated.  Thanks.

**Comment by DHowett** [comment, marked as answer]
I think your best bet would be to mimic what the `clear` command does on Linux:

Send `\e[2J\e[H`.

1. `\e[2J` is `ED` or Erase in Display, above and below the cursor
2. `\e[H` to home the cursor at the top left

Most popular terminals these days have settled on ED pushing the screen into the scrollback, so this is the most broadly compatible option!

**Reply by DHowett** [reply]
Further, I *believe* that after <kbd>Ctrl+L</kbd>, bash and tcsh just re-print the prompt plus the partial edit line. It’s not ideal, but it works!

**Comment by vefatica** [comment]
As I read the docs, ESC[2J will "Replace all text in the current viewport/screen specified by <n> with space characters".  Won't that wipe out anything that's already in the viewport?

**Reply by DHowett** [reply]
> Most popular terminals these days have settled on ED pushing the screen into the scrollback, so this is the most broadly compatible option!

Yes, it is specified as replacing all text in the current viewport with space characters. However, most popular terminals (WT included) have settled on implementing it by scrolling the contents of the screen up off the top, into the scrollback.

**Comment by vefatica** [comment]
P.S.  How do I navigate to discussions?  Edit: Never mind.  That's easy enough.

**Comment by vefatica** [comment]
Hmmm!  I see it does do that.  Maybe the docs could reflect the actual behavior.  And this one

![image](https://user-images.githubusercontent.com/61856645/176915602-1f988ffc-b4c5-429b-bb3d-048a70760bd8.png)

which mentions only scrolling (panning) clobbers previous outout.  Below  is before and after panning the viewport down (and scrolling the text up a little to see what's been clobbered).  _ROW is the row of the current command_line_in_progress and I'm moving the cursor too (which is necessary).

![image](https://user-images.githubusercontent.com/61856645/176918251-fe06bbe6-9804-42ac-b579-27ff214f5f5e.png)

![image](https://user-images.githubusercontent.com/61856645/176917667-91fdfdd0-d4ba-4d5d-b5d4-c0e142ef015c.png)

If that's correct behavior, the docs could use a little work there also.
